server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java

```diff
@@ -282,6 +531,8 @@ public void onNewClusterState(ClusterState state) {
 
             @Override
             public void onClusterServiceClose() {
+                // TODO: There is likely very little we can do about this, and our time in this JVM is likely short,
+                // should we just get out of here ASAP?
```

Comment:
So, this affects regular indexing, but if it reaches this point for regular indexing the entire operation fails and we move on. If this happens during the failure store indexing section, then we may still have accepted some of the documents in the request. It's a question of whether or not it is the right move to simply reject the failure store parts and return the request with the failures or if it is better to just fail the entire request since we may be racing with the JVM shutting down.

Or at least I assume we're racing with a JVM that is shutting down, I haven't explored the intricacies of why else the cluster service might be closed.

Reply:
Even if there are other reasons for the cluster service shutting down, it's safe to assume that multiple things will not be working anymore and they are not coming back. So, I think to start with we could fail the request.

Reply:
I have removed the TODO marker but left most of the comment intact. In the case that this is the regular indexing, we continue with the logic as it existed before: We immediately fail the request before doing any work. In the event that the cluster closes while we are about to send the failure information, then we once again will simply fail the entire request. The second phase will not be successful anyway, and we should free up the threads without any delay to allow for a timely shutdown.